pkg/manager/internal/controllers/projectroletemplate_controller.go

```diff
@@ -0,0 +1,268 @@
+			Type:    corev1alpha1.ProjectRoleTemplateReady,
+			Status:  corev1alpha1.ConditionFalse,
+			Reason:  corev1alpha1.ProjectRoleTemplateTerminatingReason,
+			Message: "ProjectRoleTemplate is being terminated",
+		})
+		if err := r.Status().Update(ctx, projectRoleTemplate); err != nil {
+			return fmt.Errorf("updating ProjectRoleTemplate status: %w", err)
+		}
+	}
+
+	cleanedUp, err := util.DeleteObjects(ctx, r.Client, r.Scheme, []runtime.Object{
+		&rbacv1.Role{},
+		&rbacv1.RoleBinding{},
+	}, owner.OwnedBy(projectRoleTemplate, r.Scheme))
+	if err != nil {
+		return fmt.Errorf("DeleteObjects: %w", err)
+	}
+	if cleanedUp && util.RemoveFinalizer(projectRoleTemplate, projectRoleTemplateControllerFinalizer) {
+		if err := r.Update(ctx, projectRoleTemplate); err != nil {
+			return fmt.Errorf("updating ProjectRoleTemplate Status: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *ProjectRoleTemplateReconciler) reconcileRBACForProject(ctx context.Context, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate, project *storagev1alpha1.Project) error {
+	// Reconcile Role.
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      projectRoleTemplate.Name,
+			Namespace: project.Status.Namespace.Name,
+		},
+		Rules: projectRoleTemplate.Spec.Rules,
+	}
+	if err := r.reconcileRole(ctx, role, projectRoleTemplate); err != nil {
+		return err
+	}
+
+	// Reconcile RoleBindings.
+	if projectRoleTemplate.HasBinding(corev1alpha1.BindToEveryone) {
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      projectRoleTemplate.Name,
+				Namespace: project.Status.Namespace.Name,
+			},
+			Subjects: project.Status.Members,
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "Role",
+				Name:     role.Name,
+			},
+		}
+		if err := r.reconcileRoleBinding(ctx, roleBinding, projectRoleTemplate); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ProjectRoleTemplateReconciler) reconcileRole(ctx context.Context, role *rbacv1.Role, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate) error {
+	if _, err := owner.ReconcileOwnedObjects(ctx, r.Client, r.Log, r.Scheme,
+		projectRoleTemplate,
+		[]runtime.Object{role}, &rbacv1.Role{},
+		func(actual, desired runtime.Object) error {
+			actualRule := actual.(*rbacv1.Role)
+			desiredRole := desired.(*rbacv1.Role)
+			if !reflect.DeepEqual(actualRule.Rules, desiredRole.Rules) {
+				actualRule.Rules = desiredRole.Rules
+			}
+			return nil
+		}); err != nil {
+		return fmt.Errorf("cannot reconcile Role: %w", err)
+	}
+	return nil
+}
+
+func (r *ProjectRoleTemplateReconciler) reconcileRoleBinding(ctx context.Context, roleBinding *rbacv1.RoleBinding, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate) error {
+	if _, err := owner.ReconcileOwnedObjects(ctx, r.Client, r.Log, r.Scheme,
+		projectRoleTemplate,
+		[]runtime.Object{roleBinding}, &rbacv1.RoleBinding{},
+		func(actual, desired runtime.Object) error {
```

Comment:
This function is poor men's server-side apply; maybe extract this function in separate package? This applies to other functions as well.

Reply:
I just noticed that `owner.ReconcileOwnedObjects` calls `controllerutil.CreateOrUpdate` underneath, so there is no need to do `reflect.DeepEqual` since `DeepEqual` will be performed in `controllerutil.CreateOrUpdate`https://github.com/kubernetes-sigs/controller-runtime/blob/4c7b43ac225cc2e6e983d1a9f9f9db4ad15c8656/pkg/controller/controllerutil/controllerutil.go#L216